### PR TITLE
Add draft of spec

### PR DESCRIPTION
**Pull request by mreichhoff**
This ensures that CORS + `use-credentials` is used in top-level subresource requests, and leaves open the possibility of an rSA call indicating embeddee opt-in for frames (with the latter point to be clarified in [a later effort](https://github.com/mreichhoff/requestStorageAccessForOrigin/issues/8) based on the outcome of per-frame discussions elsewhere).

### CODE REVIEW

**Comment by johannhof** [comment]
Ah, didn't notice this when looking at #5 😆. Similar comments probably apply :)

**Comment by mreichhoff** [comment]
yup, I addressed your comments in this one 😄 

I think we're pretty clearly moving to "require explicit rSA invocation" at this point, so this is the better PR. It's just a matter of clarifying the subresource request portion.

**Comment by mreichhoff** [comment]
got one round of review from @johannhof already...wondering if @domfarolino would want to take a look as well (we discussed parts of this today out-of-band). There's probably a good bit of work still needed, but the more we can knock out now, the better.

File: requestStorageAccessForOrigin.bs
Comment (domfarolino) [inline requestStorageAccessForOrigin.bs:25]: Isn't this exported in https://github.com/whatwg/url/blob/main/url.bs#L2721?

File: requestStorageAccessForOrigin.bs
Comment (domfarolino) [inline requestStorageAccessForOrigin.bs:29]: Same here: https://github.com/whatwg/infra/blob/main/infra.bs#L120

File: requestStorageAccessForOrigin.bs
Comment (domfarolino) [inline requestStorageAccessForOrigin.bs:37]: Exported in https://github.com/whatwg/fetch/blob/main/fetch.bs#L1504

File: requestStorageAccessForOrigin.bs
Comment (domfarolino) [inline requestStorageAccessForOrigin.bs:49]: I'd lean towards using "navigable" instead of "browsing context" these days. Specifically, I'd use "traversable navigable" here instead of top-level browsing. Your options are "traversable navigable" and "top-level traversable" here, and I would use the former which is a superset of the latter. And I'd also link to it: https://html.spec.whatwg.org/multipage/document-sequences.html#traversable-navigable.

The reason I recommend using traversable navigable over top-level traversable is because you probably want to treat i.e., fenced frames and portals as "top-level" in this context. Those frames are rooted at a traversable navigable with a non-null parent.

File: requestStorageAccessForOrigin.bs
Comment (domfarolino) [inline requestStorageAccessForOrigin.bs:60]: Out of curiosity, is there a reason you're linking to URL's definition of "origin" here and throughout this spec, as opposed to [HTML's](https://html.spec.whatwg.org/multipage/browsers.html#concept-origin)? I think most things link to HTML, since it describes what an origin is (as opposed to how to derive one from a URL, which is not the only way to obtain an origin btw), which I think is what you want.

File: requestStorageAccessForOrigin.bs
Comment (domfarolino) [inline requestStorageAccessForOrigin.bs:90]: nit: prefer to put things in terms of navigable these days. Specifically, I'd use https://html.spec.whatwg.org/#traversable-navigable which is exported.

So: "If document's [node navigable](https://html.spec.whatwg.org/#node-navigable) is not a [traversable navigable](https://html.spec.whatwg.org/#traversable-navigable) [...]"

File: requestStorageAccessForOrigin.bs
Comment (domfarolino) [inline requestStorageAccessForOrigin.bs:96]: Do we need the "/" in "resolve"? We didn't in "[=reject=]" above?

File: requestStorageAccessForOrigin.bs
Comment (domfarolino) [inline requestStorageAccessForOrigin.bs:101]: nit: Use oxford comma: https://whatwg.org/style-guide#grammar. Not sure if W3C has a different style guide but IMO we should adhere to WHATWG, which is where this spec will end up in the future if the stars align

File: requestStorageAccessForOrigin.bs
Comment (domfarolino) [inline requestStorageAccessForOrigin.bs:102]: The result of |hasAccess| is just a promise right? Do you mean to wait for the result of |hasAccess| and resolve/reject |p| with the same value? It seems like "Determine the top-level storage access policy" already queues its own task on the global it receives, to resolve/reject the promise it is given though (which is |hasAccess| IIUC), so I'm a bit confused what is going on here. It's also a little late so maybe I'm just not squinting hard enough :)

File: requestStorageAccessForOrigin.bs
Comment (domfarolino) [inline requestStorageAccessForOrigin.bs:119]: It would be good to know where this is supposed to be running? On the "main thread", or "in parallel"? It seems to kind of interact with main thread state, but earlier it seems to be called in parallel.

**Review by domfarolino — COMMENTED** [review]
Took a quick pass, can finish the rest next week but here's a few comments for now.

Btw, feel free to use https://github.com/domfarolino/specfmt if you feel so daring to format the spec incrementally and be one of that project's first guinea pigs :)

File: requestStorageAccessForOrigin.bs
Comment (mreichhoff) [reply requestStorageAccessForOrigin.bs:60]: it's just due to not knowing this distinction; switching

(side note: there seem to be conflicting `origin` exports, as I get a message like:
`The following refs show up multiple times in their spec, in a way that Bikeshed can't distinguish between. Either create a manual link, or ask the spec maintainer to add disambiguating attributes` when adding the line to `link-defaults` that it asked me for when I switched to the HTML version. 🤷

File: requestStorageAccessForOrigin.bs
Comment (mreichhoff) [reply requestStorageAccessForOrigin.bs:99]: ah, the former (again modeled on the storage access API spec, perhaps poorly).

I modified the other function to just resolve or reject the promise it is given, which might help?

File: requestStorageAccessForOrigin.bs
Comment (mreichhoff) [reply requestStorageAccessForOrigin.bs]: (removed the `global` variable and the top-level site in favor of plumbing it in; attempted to follow [the docs](https://html.spec.whatwg.org/multipage/infrastructure.html#in-parallel), though I suspect user activation will also have to be plumbed in)

File: requestStorageAccessForOrigin.bs
Comment (mreichhoff) [reply requestStorageAccessForOrigin.bs]: switched to that in the latest version

File: requestStorageAccessForOrigin.bs
Comment (domfarolino) [inline requestStorageAccessForOrigin.bs:111]: This variable is of type "URL" but the variable name is "origin". That's a little confusing, but I guess it matches the fact that TopLevelStorageAccessPermissionDescriptor's requestedOrigin is a USVString (not an origin), but I think we still need to pull the _serialized_ origin out of the URL right? Or else the dictionary's requestedOrigin member will actually be the entire URL string?

Comment (mreichhoff) [reply requestStorageAccessForOrigin.bs]: hm...I went with USVString for the descriptor, but perhaps it would be better to use origin? I wasn't sure what the correct IDL type would be. Regardless, yes, I at least switched to the request's url's origin here.

File: requestStorageAccessForOrigin.bs
Comment (mreichhoff) [reply requestStorageAccessForOrigin.bs]: done

File: requestStorageAccessForOrigin.bs
Comment (mreichhoff) [reply requestStorageAccessForOrigin.bs]: done

File: requestStorageAccessForOrigin.bs
Comment (mreichhoff) [reply requestStorageAccessForOrigin.bs]: done

File: requestStorageAccessForOrigin.bs
Comment (mreichhoff) [reply requestStorageAccessForOrigin.bs:85]: done

File: requestStorageAccessForOrigin.bs
Comment (mreichhoff) [reply requestStorageAccessForOrigin.bs]: yep! Done.

File: requestStorageAccessForOrigin.bs
Comment (johannhof) [inline requestStorageAccessForOrigin.bs:107]: Can we file this issue on the repo and link there?

File: requestStorageAccessForOrigin.bs
Comment (johannhof) [inline requestStorageAccessForOrigin.bs:122]: Is this TODO still open?

**Review by johannhof — APPROVED** [review]
This seems good to merge from my side :)

File: requestStorageAccessForOrigin.bs
Comment (mreichhoff) [reply requestStorageAccessForOrigin.bs]: good call, done

File: requestStorageAccessForOrigin.bs
Comment (mreichhoff) [reply requestStorageAccessForOrigin.bs]: no, removed it

File: requestStorageAccessForOrigin.bs
Comment (domfarolino) [inline requestStorageAccessForOrigin.bs:128]: This is a bit handwavy as we're doing some action-at-a-distance, looking backwards to state that lived on a different thread. I think it's best to pull the state of the user activation out early on the main thread and plumb it through

File: requestStorageAccessForOrigin.bs
Comment (domfarolino) [inline requestStorageAccessForOrigin.bs:140]: We have to queue a task to do this on the main thread right? Also nit: "Reject", since it is the beginning of a sentence

File: requestStorageAccessForOrigin.bs
Comment (domfarolino) [inline requestStorageAccessForOrigin.bs:146]: User activation querying and consumption should be done on the main thread. How does this work implementation-wise?

File: requestStorageAccessForOrigin.bs
Comment (domfarolino) [inline requestStorageAccessForOrigin.bs:108]: Should "/origin" be "url/origin" here?

**Review by domfarolino — COMMENTED** [review]
A few nits mostly around touching main-thread state from an in-parallel context. This is probably good enough for a draft provided we follow-up on some of the comments accordingly. It'll probably be another week until I can take another close look, so feel free to not block on me if you need to publish. I can take another close look after the holidays or maybe in between them etc.

**Comment by mreichhoff** [comment]
will address remaining domfarolino@ comments in a follow-up PR (cleaning this up as it was from before moving the repo). Thanks!

**Pull request merged**
